# Transforming unstructured data (JSON) to structured data (CSV)

Loading the Users, Brands and Receipts data into pandas dataframe and converting them to CSV format

In [1]:
import pandas as pd
#Loading the Users data into pandas dataframe
users_df = pd.read_json("/Users/mikeakshaydominique/Desktop/users.json", lines = True)

In [2]:
#Converting to CSV
users_df.to_csv("users.csv", index=False)

In [3]:
print(users_df)

                                      _id  active               createdDate  \
0    {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
1    {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
2    {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
3    {'$oid': '5ff1e1eacfcf6c399c274ae6'}    True  {'$date': 1609687530554}   
4    {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
..                                    ...     ...                       ...   
490  {'$oid': '54943462e4b07e684157a532'}    True  {'$date': 1418998882381}   
491  {'$oid': '54943462e4b07e684157a532'}    True  {'$date': 1418998882381}   
492  {'$oid': '54943462e4b07e684157a532'}    True  {'$date': 1418998882381}   
493  {'$oid': '54943462e4b07e684157a532'}    True  {'$date': 1418998882381}   
494  {'$oid': '54943462e4b07e684157a532'}    True  {'$date': 1418998882381}   

                    lastLogin         role signUpSo

In [4]:
#Loading the Brands data into pandas dataframe
brands_df = pd.read_json("/Users/mikeakshaydominique/Desktop/brands.json", lines = True)

In [5]:
#Converting to CSV
brands_df.to_csv("brands.csv", index=False)

In [6]:
print(brands_df)

                                       _id       barcode            category  \
0     {'$oid': '601ac115be37ce2ead437551'}  511111019862              Baking   
1     {'$oid': '601c5460be37ce2ead43755f'}  511111519928           Beverages   
2     {'$oid': '601ac142be37ce2ead43755d'}  511111819905              Baking   
3     {'$oid': '601ac142be37ce2ead43755a'}  511111519874              Baking   
4     {'$oid': '601ac142be37ce2ead43755e'}  511111319917      Candy & Sweets   
...                                    ...           ...                 ...   
1162  {'$oid': '5f77274dbe37ce6b592e90c0'}  511111116752              Baking   
1163  {'$oid': '5dc1fca91dda2c0ad7da64ae'}  511111706328  Breakfast & Cereal   
1164  {'$oid': '5f494c6e04db711dd8fe87e7'}  511111416173      Candy & Sweets   
1165  {'$oid': '5a021611e4b00efe02b02a57'}  511111400608             Grocery   
1166  {'$oid': '6026d757be37ce6369301468'}  511111019930              Baking   

          categoryCode                 

In [7]:
#Loading the Receipts data into pandas dataframe
receipts_df = pd.read_json("/Users/mikeakshaydominique/Desktop/receipts.json", lines = True)

In [8]:
#Converting to CSV
receipts_df.to_csv("receipts.csv", index=False)

In [9]:
print(receipts_df)

                                       _id  bonusPointsEarned  \
0     {'$oid': '5ff1e1eb0a720f0523000575'}              500.0   
1     {'$oid': '5ff1e1bb0a720f052300056b'}              150.0   
2     {'$oid': '5ff1e1f10a720f052300057a'}                5.0   
3     {'$oid': '5ff1e1ee0a7214ada100056f'}                5.0   
4     {'$oid': '5ff1e1d20a7214ada1000561'}                5.0   
...                                    ...                ...   
1114  {'$oid': '603cc0630a720fde100003e6'}               25.0   
1115  {'$oid': '603d0b710a720fde1000042a'}                NaN   
1116  {'$oid': '603cf5290a720fde10000413'}                NaN   
1117  {'$oid': '603ce7100a7217c72c000405'}               25.0   
1118  {'$oid': '603c4fea0a7217c72c000389'}                NaN   

                                bonusPointsEarnedReason  \
0     Receipt number 2 completed, bonus point schedu...   
1     Receipt number 5 completed, bonus point schedu...   
2                            All-receipts 

# Data Cleaning (Users data)

Checking if there are any missing values in the 'Users' data

In [10]:
if users_df.isnull().any().any():
    print("There are missing values in the Users DataFrame.")
else:
    print("There are no missing values in the Users DataFrame.")

There are missing values in the Users DataFrame.


In [11]:
#Creating a boolean mask to identify missing values
missing_mask = users_df.isnull()

In [12]:
# Total number of missing values in the entire DataFrame
total_missing = missing_mask.sum().sum()
print("Total missing values in the DataFrame:", total_missing)

Total missing values in the DataFrame: 166


In [15]:
# Counting the number of missing values in each column
missing_values_in_columns = users_df.isnull().sum()
print(missing_values_in_columns )

# Counting the number of missing values in each row
missing_values_in_rows = users_df.isnull().sum(axis=1)
print(missing_values_in_rows)

_id              0
active           0
createdDate      0
lastLogin       62
role             0
signUpSource    48
state           56
dtype: int64
0      0
1      0
2      0
3      0
4      0
      ..
490    2
491    2
492    2
493    2
494    2
Length: 495, dtype: int64


Handling the 48 missing values found in the 'signUpSource' column

In [17]:
# Filling missing values in the 'signUpSource' column with 'Unknown'
users_df['signUpSource'].fillna('Unknown', inplace=True)

In [43]:
print(users_df['signUpSource'])

0        Email
1        Email
2        Email
3        Email
4        Email
        ...   
490    Unknown
491    Unknown
492    Unknown
493    Unknown
494    Unknown
Name: signUpSource, Length: 495, dtype: object


In [19]:
#Cross-verifying the number of missing values in the column to ensure the missing values are handled successfully 
missing_values_count = users_df['signUpSource'].isnull().sum()
print("Number of missing values in 'signUpSource':", missing_values_count)

Number of missing values in 'signUpSource': 0


Likewise, the 56 missing values values in the 'state' column are also handled.

In [20]:
# Filling missing values in the 'state' column with 'Unknown'
users_df['state'].fillna('Unknown', inplace=True)

In [41]:
print(users_df['state'])

0           WI
1           WI
2           WI
3           WI
4           WI
        ...   
490    Unknown
491    Unknown
492    Unknown
493    Unknown
494    Unknown
Name: state, Length: 495, dtype: object


In [22]:
#Cross-verifying the number of missing values in the column to ensure the missing values are handled successfully 
missing_values_count = users_df['state'].isnull().sum()
print("Number of missing values in 'state':", missing_values_count)

Number of missing values in 'state': 0


Converting the values in the columns 'lastLogin' and 'createdDate' from the given format to a usable datetime format.

In [23]:
# Defining a custom function to extract the timestamp from the nested dictionary
def extract_timestamp(val):
    if pd.notna(val):
        return val['$date']
    else:
        return val

# Converting the values in 'lastLogin' to a datetime format using the custom function
users_df['lastLogin'] = users_df['lastLogin'].apply(extract_timestamp)
users_df['lastLogin'] = pd.to_datetime(users_df['lastLogin'], unit='ms')

# Converting the values in 'createdDate' to a datetime format using the custom function
users_df['createdDate'] = users_df['createdDate'].apply(extract_timestamp)
users_df['createdDate'] = pd.to_datetime(users_df['createdDate'], unit='ms')

In [202]:
print(users_df['lastLogin'], users_df['createdDate'])

0     2021-01-03 15:25:37.857999872
1     2021-01-03 15:25:37.857999872
2     2021-01-03 15:25:37.857999872
3     2021-01-03 15:25:30.596999936
4     2021-01-03 15:25:37.857999872
                   ...             
490   2021-03-05 16:52:23.204000000
491   2021-03-05 16:52:23.204000000
492   2021-03-05 16:52:23.204000000
493   2021-03-05 16:52:23.204000000
494   2021-03-05 16:52:23.204000000
Name: lastLogin, Length: 495, dtype: datetime64[ns] 0     2021-01-03 15:24:04.800
1     2021-01-03 15:24:04.800
2     2021-01-03 15:24:04.800
3     2021-01-03 15:25:30.554
4     2021-01-03 15:24:04.800
                ...          
490   2014-12-19 14:21:22.381
491   2014-12-19 14:21:22.381
492   2014-12-19 14:21:22.381
493   2014-12-19 14:21:22.381
494   2014-12-19 14:21:22.381
Name: createdDate, Length: 495, dtype: datetime64[ns]


In [26]:
missing_values_count = users_df['lastLogin'].isnull().sum()
print("Number of missing values in 'lastLogin':", missing_values_count)

missing_values_count = users_df['createdDate'].isnull().sum()
print("Number of missing values in 'createdDate':", missing_values_count)

Number of missing values in 'lastLogin': 62
Number of missing values in 'createdDate': 0


Handling the missing values in the 'lastLogin' column.

In [27]:
#Forward-filling missing values in 'lastLogin' column
users_df['lastLogin'].fillna(method='ffill', inplace=True)

In [28]:
missing_values_count = users_df['lastLogin'].isnull().sum()
print("Number of missing values in 'lastLogin':", missing_values_count)

Number of missing values in 'lastLogin': 0


Converting the '_id' column into a more useable format

In [30]:
# Defining a custom function to extract the BSON ObjectID from the dictionary
def extract_oid(val):
    if pd.notna(val):  # Checking for NaN or missing values
        return val['$oid']
    else:
        return val

# Extracting the BSON ObjectID from the dictionary
users_df['_id'] = users_df['_id'].apply(extract_oid)

In [31]:
print(users_df['_id'])

0      5ff1e194b6a9d73a3a9f1052
1      5ff1e194b6a9d73a3a9f1052
2      5ff1e194b6a9d73a3a9f1052
3      5ff1e1eacfcf6c399c274ae6
4      5ff1e194b6a9d73a3a9f1052
                 ...           
490    54943462e4b07e684157a532
491    54943462e4b07e684157a532
492    54943462e4b07e684157a532
493    54943462e4b07e684157a532
494    54943462e4b07e684157a532
Name: _id, Length: 495, dtype: object


# Data Cleaning (Brands data)

Checking if there are any missing values in the 'Brands' data

In [32]:
if brands_df.isnull().any().any():
    print("There are missing values in the Brands DataFrame.")
else:
    print("There are no missing values in the Brands DataFrame.")

There are missing values in the Brands DataFrame.


In [33]:
#Creating a boolean mask to identify missing values
missing_mask_brands = brands_df.isnull()

In [34]:
# Total number of missing values in the entire DataFrame
total_missing_values = missing_mask_brands.sum().sum()
print("Total missing values in the DataFrame:", total_missing_values)

Total missing values in the DataFrame: 1651


In [37]:
# Counting the number of missing values in each column
missing_values_in_columns_brands = brands_df.isnull().sum()
print(missing_values_in_columns_brands)

# Counting the number of missing values in each row
missing_values_in_rows_brands = brands_df.isnull().sum(axis=1)
print(missing_values_in_rows_brands)

_id               0
barcode           0
category        155
categoryCode    650
cpg               0
name              0
topBrand        612
brandCode       234
dtype: int64
0       1
1       0
2       0
3       0
4       0
       ..
1162    2
1163    2
1164    1
1165    1
1166    0
Length: 1167, dtype: int64


Handling the 155 missing values found in the 'category' column

In [47]:
# Filling missing values in the 'category' column with 'Unknown'
brands_df['category'].fillna('Unknown', inplace=True)

In [48]:
print(brands_df['category'])

0                   Baking
1                Beverages
2                   Baking
3                   Baking
4           Candy & Sweets
               ...        
1162                Baking
1163    Breakfast & Cereal
1164        Candy & Sweets
1165               Grocery
1166                Baking
Name: category, Length: 1167, dtype: object


In [45]:
#Cross-verifying the number of missing values in the column to ensure the missing values are handled successfully 
missing_values_count_ = brands_df['category'].isnull().sum()
print("Number of missing values in 'category':", missing_values_count_)

Number of missing values in 'category': 0


Likewise, the 650 missing values values in the 'categoryCode' column are handled.

In [49]:
# Filling missing values in the 'categoryCode' column with 'Unknown'
brands_df['categoryCode'].fillna('Unknown', inplace=True)

In [50]:
print(brands_df['categoryCode'])

0                 BAKING
1              BEVERAGES
2                 BAKING
3                 BAKING
4       CANDY_AND_SWEETS
              ...       
1162              BAKING
1163             Unknown
1164    CANDY_AND_SWEETS
1165             Unknown
1166              BAKING
Name: categoryCode, Length: 1167, dtype: object


In [52]:
#Cross-verifying the number of missing values in the column to ensure the missing values are handled successfully 
missing_val_count_ = brands_df['categoryCode'].isnull().sum()
print("Number of missing values in 'categoryCode':", missing_val_count_)

Number of missing values in 'categoryCode': 0


Handling the 612 missing values in the 'topBrand' column

In [54]:
# Checking the data type of the 'topBrand' column
data_type = brands_df['topBrand'].dtype

if data_type == 'object':
    # The column is categorical (represented as an object/string data type)
    print("The 'topBrand' column is categorical.")
else:
    # The column is not categorical
    print("The 'topBrand' column is not categorical.")

The 'topBrand' column is not categorical.


In [55]:
# Filling missing values with 0.0
brands_df['topBrand'].fillna(0.0, inplace=True)

In [56]:
print(brands_df['topBrand'])

0       0.0
1       0.0
2       0.0
3       0.0
4       0.0
       ... 
1162    0.0
1163    0.0
1164    0.0
1165    0.0
1166    0.0
Name: topBrand, Length: 1167, dtype: float64


In [57]:
#Cross-verifying the number of missing values in the column to ensure the missing values are handled successfully 
missing_val_count = brands_df['topBrand'].isnull().sum()
print("Number of missing values in 'topBrand':", missing_val_count)

Number of missing values in 'topBrand': 0


In [58]:
print(brands_df['brandCode'])

0                                 NaN
1                           STARBUCKS
2       TEST BRANDCODE @1612366146176
3       TEST BRANDCODE @1612366146051
4       TEST BRANDCODE @1612366146827
                    ...              
1162                              NaN
1163               DIPPIN DOTS CEREAL
1164    TEST BRANDCODE @1598639215217
1165                LIPTON TEA Leaves
1166    TEST BRANDCODE @1613158231644
Name: brandCode, Length: 1167, dtype: object


In [59]:
# Filling NaN and test codes with 'Unknown'
brands_df['brandCode'].fillna('Unknown', inplace=True)
brands_df['brandCode'] = brands_df['brandCode'].str.replace(r'^TEST BRANDCODE @\d+$', 'Unknown', regex=True)

In [60]:
print(brands_df['brandCode'])

0                  Unknown
1                STARBUCKS
2                  Unknown
3                  Unknown
4                  Unknown
               ...        
1162               Unknown
1163    DIPPIN DOTS CEREAL
1164               Unknown
1165     LIPTON TEA Leaves
1166               Unknown
Name: brandCode, Length: 1167, dtype: object


In [61]:
#Cross-verifying the number of missing values in the column to ensure the missing values are handled successfully 
missing_val_cnt_ = brands_df['brandCode'].isnull().sum()
print("Number of missing values in 'brandCode':", missing_val_cnt_)

Number of missing values in 'brandCode': 0


Extracting the time stamps from the nested dictionary in the '_id' column and storing it as strings

In [174]:
# Defining a custom function to extract the timestamp from the nested dictionary
def extract_timestamp(val):
    if pd.notna(val):
        return val['$oid']
    else:
        return val

# Storing the extracted time stamps as strings in the _id column
brands_df['_id'] = brands_df['_id'].apply(extract_timestamp)

In [175]:
print(brands_df['_id'])

0       601ac115be37ce2ead437551
1       601c5460be37ce2ead43755f
2       601ac142be37ce2ead43755d
3       601ac142be37ce2ead43755a
4       601ac142be37ce2ead43755e
                  ...           
1162    5f77274dbe37ce6b592e90c0
1163    5dc1fca91dda2c0ad7da64ae
1164    5f494c6e04db711dd8fe87e7
1165    5a021611e4b00efe02b02a57
1166    6026d757be37ce6369301468
Name: _id, Length: 1167, dtype: object


Cleaning the 'cpg' column

In [176]:
# Defining a custom function to extract the relevant information from the 'cpg' column
def extract_cpg_info(cpg_data):
    if isinstance(cpg_data, dict):
        # If the 'cpg_data' is a dictionary, extract the values for each key
        return cpg_data.get('$id'), cpg_data.get('$ref')
    else:
        # If the 'cpg_data' is a reference, split and extract the reference value
        return cpg_data.split(',')[1].split(':')[1].strip()

# Applying the custom function to the 'cpg' column and creating new columns for the extracted data
brands_df[['cpg_id', 'cpg_ref']] = brands_df['cpg'].apply(extract_cpg_info).apply(pd.Series)

In [181]:
# Dropping the original 'cpg' column
brands_df.drop(columns=['cpg'], inplace=True)

In [178]:
print(brands_df['cpg_id'])

0       {'$oid': '601ac114be37ce2ead437550'}
1       {'$oid': '5332f5fbe4b03c9a25efd0ba'}
2       {'$oid': '601ac142be37ce2ead437559'}
3       {'$oid': '601ac142be37ce2ead437559'}
4       {'$oid': '5332fa12e4b03c9a25efd1e7'}
                        ...                 
1162    {'$oid': '5f77274dbe37ce6b592e90bf'}
1163    {'$oid': '53e10d6368abd3c7065097cc'}
1164    {'$oid': '5332fa12e4b03c9a25efd1e7'}
1165    {'$oid': '5332f5f6e4b03c9a25efd0b4'}
1166    {'$oid': '6026d757be37ce6369301467'}
Name: cpg_id, Length: 1167, dtype: object


Extracting the time stamps from the nested dictionary in the 'cpg_id' column and storing it as strings

In [188]:
# Defining a custom function to extract the timestamp from the nested dictionary
def extract_timestamp(val):
    if pd.notna(val):
        return val['$oid']
    else:
        return val

# Storing the extracted time stamps as strings in the cpg_id column
brands_df['cpg_id'] = brands_df['cpg_id'].apply(extract_timestamp)

In [189]:
print(brands_df['cpg_id'])

0       601ac114be37ce2ead437550
1       5332f5fbe4b03c9a25efd0ba
2       601ac142be37ce2ead437559
3       601ac142be37ce2ead437559
4       5332fa12e4b03c9a25efd1e7
                  ...           
1162    5f77274dbe37ce6b592e90bf
1163    53e10d6368abd3c7065097cc
1164    5332fa12e4b03c9a25efd1e7
1165    5332f5f6e4b03c9a25efd0b4
1166    6026d757be37ce6369301467
Name: cpg_id, Length: 1167, dtype: object


In [272]:
brands_df

_id       barcode            category  \
0     601ac115be37ce2ead437551  511111019862              Baking   
1     601c5460be37ce2ead43755f  511111519928           Beverages   
2     601ac142be37ce2ead43755d  511111819905              Baking   
3     601ac142be37ce2ead43755a  511111519874              Baking   
4     601ac142be37ce2ead43755e  511111319917      Candy & Sweets   
...                        ...           ...                 ...   
1162  5f77274dbe37ce6b592e90c0  511111116752              Baking   
1163  5dc1fca91dda2c0ad7da64ae  511111706328  Breakfast & Cereal   
1164  5f494c6e04db711dd8fe87e7  511111416173      Candy & Sweets   
1165  5a021611e4b00efe02b02a57  511111400608             Grocery   
1166  6026d757be37ce6369301468  511111019930              Baking   

          categoryCode                       name  topBrand  \
0               BAKING  test brand @1612366101024       0.0   
1            BEVERAGES                  Starbucks       0.0   
2               BAKING  test brand @1612366146176       0.0   
3               BAKING  test brand @1612366146051       0.0   
4     CANDY_AND_SWEETS  test brand @1612366146827       0.0   
...                ...                        ...       ...   
1162            BAKING  test brand @1601644365844       0.0   
1163           Unknown        Dippin Dots® Cereal       0.0   
1164  CANDY_AND_SWEETS  test brand @1598639215217       0.0   
1165           Unknown          LIPTON TEA Leaves       0.0   
1166            BAKING  test brand @1613158231643       0.0   

               brandCode                    cpg_id cpg_ref  
0                Unknown  601ac114be37ce2ead437550    Cogs  
1              STARBUCKS  5332f5fbe4b03c9a25efd0ba    Cogs  
2                Unknown  601ac142be37ce2ead437559    Cogs  
3                Unknown  601ac142be37ce2ead437559    Cogs  
4                Unknown  5332fa12e4b03c9a25efd1e7    Cogs  
...                  ...                       ...     ...  
1162             Unknown  5f77274dbe37ce6b592e90bf    Cogs  
1163  DIPPIN DOTS CEREAL  53e10d6368abd3c7065097cc    Cogs  
1164             Unknown  5332fa12e4b03c9a25efd1e7    Cogs  
1165   LIPTON TEA Leaves  5332f5f6e4b03c9a25efd0b4    Cogs  
1166             Unknown  6026d757be37ce6369301467    Cogs  

[1167 rows x 9 columns]

# Data Cleaning (Receipts data)

When checked, it was found that the 'Receipts' data had missing values as well.

In [63]:
#Creating a boolean mask to identify missing values
missing_mask_receipts = receipts_df.isnull()

In [64]:
# Total number of missing values in the entire DataFrame
total_missing_val = missing_mask_receipts.sum().sum()
print("Total missing values in the DataFrame:", total_missing_val)

Total missing values in the DataFrame: 4600


In [66]:
# Counting the number of missing values in each column
missing_values_in_columns_receipts = receipts_df.isnull().sum()
print(missing_values_in_columns_receipts)

# Counting the number of missing values in each row
missing_values_in_rows_receipts = receipts_df.isnull().sum(axis=1)
print(missing_values_in_rows_receipts)

_id                          0
bonusPointsEarned          575
bonusPointsEarnedReason    575
createDate                   0
dateScanned                  0
finishedDate               551
modifyDate                   0
pointsAwardedDate          582
pointsEarned               510
purchaseDate               448
purchasedItemCount         484
rewardsReceiptItemList     440
rewardsReceiptStatus         0
totalSpent                 435
userId                       0
dtype: int64
0       0
1       0
2       2
3       0
4       0
       ..
1114    2
1115    9
1116    9
1117    2
1118    9
Length: 1119, dtype: int64


Handling the 575 missing values in the 'bonusPointsEarned' column.

In [67]:
print(receipts_df['bonusPointsEarned'])

0       500.0
1       150.0
2         5.0
3         5.0
4         5.0
        ...  
1114     25.0
1115      NaN
1116      NaN
1117     25.0
1118      NaN
Name: bonusPointsEarned, Length: 1119, dtype: float64


In [68]:
# Filling missing values with 0.0
receipts_df['bonusPointsEarned'].fillna(0.0, inplace=True)

In [69]:
print(receipts_df['bonusPointsEarned'])

0       500.0
1       150.0
2         5.0
3         5.0
4         5.0
        ...  
1114     25.0
1115      0.0
1116      0.0
1117     25.0
1118      0.0
Name: bonusPointsEarned, Length: 1119, dtype: float64


In [71]:
#Cross-verifying the number of missing values in the column to ensure the missing values are handled successfully 
missing_val_cnt = receipts_df['bonusPointsEarned'].isnull().sum()
print("Number of missing values in 'bonusPointsEarned':", missing_val_cnt)

Number of missing values in 'bonusPointsEarned': 0


Handling the 575 missing values in the 'bonusPointsEarnedReason' column.

In [72]:
print(receipts_df['bonusPointsEarnedReason'])

0       Receipt number 2 completed, bonus point schedu...
1       Receipt number 5 completed, bonus point schedu...
2                              All-receipts receipt bonus
3                              All-receipts receipt bonus
4                              All-receipts receipt bonus
                              ...                        
1114                          COMPLETE_NONPARTNER_RECEIPT
1115                                                  NaN
1116                                                  NaN
1117                          COMPLETE_NONPARTNER_RECEIPT
1118                                                  NaN
Name: bonusPointsEarnedReason, Length: 1119, dtype: object


In [74]:
# Filling missing values in the 'bonusPointsEarnedReason' column with 'Unknown'
receipts_df['bonusPointsEarnedReason'].fillna('Unknown', inplace=True)

In [75]:
print(receipts_df['bonusPointsEarnedReason'])

0       Receipt number 2 completed, bonus point schedu...
1       Receipt number 5 completed, bonus point schedu...
2                              All-receipts receipt bonus
3                              All-receipts receipt bonus
4                              All-receipts receipt bonus
                              ...                        
1114                          COMPLETE_NONPARTNER_RECEIPT
1115                                              Unknown
1116                                              Unknown
1117                          COMPLETE_NONPARTNER_RECEIPT
1118                                              Unknown
Name: bonusPointsEarnedReason, Length: 1119, dtype: object


In [77]:
#Cross-verifying the number of missing values in the column to ensure the missing values are handled successfully 
missing_val_cntt = receipts_df['bonusPointsEarnedReason'].isnull().sum()
print("Number of missing values in 'bonusPointsEarnedReason':", missing_val_cntt)

Number of missing values in 'bonusPointsEarnedReason': 0


Changing the format and handling missing values in the 'finishedDate' column

In [78]:
#Printing the finishedDate column to check the format of the values
print(receipts_df['finishedDate'])

0       {'$date': 1609687531000}
1       {'$date': 1609687483000}
2                            NaN
3       {'$date': 1609687534000}
4       {'$date': 1609687511000}
                  ...           
1114                         NaN
1115                         NaN
1116                         NaN
1117                         NaN
1118                         NaN
Name: finishedDate, Length: 1119, dtype: object


Converting the values in the 'finishedDate' column from the given format to a usable datetime format.

In [79]:
# Defining a custom function to extract the timestamp from the nested dictionary
def extract_timestamp(val):
    if pd.notna(val):
        return val['$date']
    else:
        return val

# Converting the values in 'finishedDate' to a datetime format using the custom function
receipts_df['finishedDate'] = receipts_df['finishedDate'].apply(extract_timestamp)
receipts_df['finishedDate'] = pd.to_datetime(receipts_df['finishedDate'], unit='ms')

Handling the 551 missing values in the 'finishedDate' column.

In [83]:
# Imputing missing values with 'Unknown'
receipts_df['finishedDate'].fillna('Unknown', inplace=True)

In [85]:
print(receipts_df['finishedDate'])

0       2021-01-03 15:25:31
1       2021-01-03 15:24:43
2                   Unknown
3       2021-01-03 15:25:34
4       2021-01-03 15:25:11
               ...         
1114                Unknown
1115                Unknown
1116                Unknown
1117                Unknown
1118                Unknown
Name: finishedDate, Length: 1119, dtype: object


In [97]:
#Cross-verifying the number of missing values in the column to ensure the missing values are handled successfully 
the_missing_val_cnt = receipts_df['finishedDate'].isnull().sum()
print("Number of missing values in 'finishedDate':", the_missing_val_cnt)

Number of missing values in 'finishedDate': 0


Converting the values in the 'pointsAwardedDate' column from the given format to a usable datetime format.

In [86]:
print(receipts_df['pointsAwardedDate'])

0       {'$date': 1609687531000}
1       {'$date': 1609687483000}
2                            NaN
3       {'$date': 1609687534000}
4       {'$date': 1609687506000}
                  ...           
1114                         NaN
1115                         NaN
1116                         NaN
1117                         NaN
1118                         NaN
Name: pointsAwardedDate, Length: 1119, dtype: object


In [87]:
# Defining a custom function to extract the timestamp from the nested dictionary
def extract_timestamp(val):
    if pd.notna(val):
        return val['$date']
    else:
        return val

# Converting the values in 'pointsAwardedDate' to a datetime format using the custom function
receipts_df['pointsAwardedDate'] = receipts_df['pointsAwardedDate'].apply(extract_timestamp)
receipts_df['pointsAwardedDate'] = pd.to_datetime(receipts_df['pointsAwardedDate'], unit='ms')

Handling the 582 missing values in the 'pointsAwardedDate' column.

In [88]:
# Imputing missing values with 'Unknown'
receipts_df['pointsAwardedDate'].fillna('Unknown', inplace=True)

In [89]:
print(receipts_df['pointsAwardedDate'])

0       2021-01-03 15:25:31
1       2021-01-03 15:24:43
2                   Unknown
3       2021-01-03 15:25:34
4       2021-01-03 15:25:06
               ...         
1114                Unknown
1115                Unknown
1116                Unknown
1117                Unknown
1118                Unknown
Name: pointsAwardedDate, Length: 1119, dtype: object


In [96]:
#Cross-verifying the number of missing values in the column to ensure the missing values are handled successfully 
missing_vals_cntt = receipts_df['pointsAwardedDate'].isnull().sum()
print("Number of missing values in 'pointsAwardedDate':", missing_vals_cntt)

Number of missing values in 'pointsAwardedDate': 0


Handling the 510 missing values in the 'pointsEarned' column.

In [90]:
print(receipts_df['pointsEarned'])

0       500.0
1       150.0
2         5.0
3         5.0
4         5.0
        ...  
1114     25.0
1115      NaN
1116      NaN
1117     25.0
1118      NaN
Name: pointsEarned, Length: 1119, dtype: float64


In [93]:
# Filling missing values with 0.0
receipts_df['pointsEarned'].fillna(0.0, inplace=True)

In [94]:
print(receipts_df['pointsEarned'])

0       500.0
1       150.0
2         5.0
3         5.0
4         5.0
        ...  
1114     25.0
1115      0.0
1116      0.0
1117     25.0
1118      0.0
Name: pointsEarned, Length: 1119, dtype: float64


In [95]:
#Cross-verifying the number of missing values in the column to ensure the missing values are handled successfully 
missing_val_cntt_ = receipts_df['pointsEarned'].isnull().sum()
print("Number of missing values in 'pointsEarned':", missing_val_cntt_)

Number of missing values in 'pointsEarned': 0


Converting the values in the 'purchaseDate' column from the given format to a usable datetime format.

In [100]:
print(receipts_df['purchaseDate'])

0       {'$date': 1609632000000}
1       {'$date': 1609601083000}
2       {'$date': 1609632000000}
3       {'$date': 1609632000000}
4       {'$date': 1609601106000}
                  ...           
1114    {'$date': 1597622400000}
1115                         NaN
1116                         NaN
1117    {'$date': 1597622400000}
1118                         NaN
Name: purchaseDate, Length: 1119, dtype: object


In [101]:
# Defining a custom function to extract the timestamp from the nested dictionary
def extract_timestamp(val):
    if pd.notna(val):
        return val['$date']
    else:
        return val

# Converting the values in 'purchaseDate' to a datetime format using the custom function
receipts_df['purchaseDate'] = receipts_df['purchaseDate'].apply(extract_timestamp)
receipts_df['purchaseDate'] = pd.to_datetime(receipts_df['purchaseDate'], unit='ms')

Handling the 448 missing values in the 'purchaseDate' column.

In [102]:
# Imputing missing values with 'Unknown'
receipts_df['purchaseDate'].fillna('Unknown', inplace=True)

In [103]:
print(receipts_df['purchaseDate'])

0       2021-01-03 00:00:00
1       2021-01-02 15:24:43
2       2021-01-03 00:00:00
3       2021-01-03 00:00:00
4       2021-01-02 15:25:06
               ...         
1114    2020-08-17 00:00:00
1115                Unknown
1116                Unknown
1117    2020-08-17 00:00:00
1118                Unknown
Name: purchaseDate, Length: 1119, dtype: object


In [104]:
#Cross-verifying the number of missing values in the column to ensure the missing values are handled successfully 
missing_vals_count = receipts_df['purchaseDate'].isnull().sum()
print("Number of missing values in 'purchaseDate':", missing_vals_count)

Number of missing values in 'purchaseDate': 0


Handling the 484 missing values in the 'purchaseItemCount' column.

In [106]:
print(receipts_df['purchasedItemCount'])

0       5.0
1       2.0
2       1.0
3       4.0
4       2.0
       ... 
1114    2.0
1115    NaN
1116    NaN
1117    2.0
1118    NaN
Name: purchasedItemCount, Length: 1119, dtype: float64


In [107]:
# Filling missing values with 0.0
receipts_df['purchasedItemCount'].fillna(0.0, inplace=True)

In [108]:
print(receipts_df['purchasedItemCount'])

0       5.0
1       2.0
2       1.0
3       4.0
4       2.0
       ... 
1114    2.0
1115    0.0
1116    0.0
1117    2.0
1118    0.0
Name: purchasedItemCount, Length: 1119, dtype: float64


In [109]:
#Cross-verifying the number of missing values in the column to ensure the missing values are handled successfully 
missing_val_count__ = receipts_df['purchasedItemCount'].isnull().sum()
print("Number of missing values in 'purchasedItemCount':", missing_val_count__)

Number of missing values in 'purchasedItemCount': 0


Handling the 440 missing values in the 'rewardsReceiptItemList' column.

In [110]:
print(receipts_df['rewardsReceiptItemList'])

0       [{'barcode': '4011', 'description': 'ITEM NOT ...
1       [{'barcode': '4011', 'description': 'ITEM NOT ...
2       [{'needsFetchReview': False, 'partnerItemId': ...
3       [{'barcode': '4011', 'description': 'ITEM NOT ...
4       [{'barcode': '4011', 'description': 'ITEM NOT ...
                              ...                        
1114    [{'barcode': 'B076FJ92M4', 'description': 'mue...
1115                                                  NaN
1116                                                  NaN
1117    [{'barcode': 'B076FJ92M4', 'description': 'mue...
1118                                                  NaN
Name: rewardsReceiptItemList, Length: 1119, dtype: object


In [115]:
# Defining a custom function to handle NaN values in the 'rewardsReceiptItemList' column
def fill_nan_with_empty_list(val):
    if isinstance(val, list):
        return val
    else:
        return []

# Applying the custom function to fill NaN values in the 'rewardsReceiptItemList' column
receipts_df['rewardsReceiptItemList'] = receipts_df['rewardsReceiptItemList'].apply(fill_nan_with_empty_list)

In [116]:
print(receipts_df['rewardsReceiptItemList'])

0       [{'barcode': '4011', 'description': 'ITEM NOT ...
1       [{'barcode': '4011', 'description': 'ITEM NOT ...
2       [{'needsFetchReview': False, 'partnerItemId': ...
3       [{'barcode': '4011', 'description': 'ITEM NOT ...
4       [{'barcode': '4011', 'description': 'ITEM NOT ...
                              ...                        
1114    [{'barcode': 'B076FJ92M4', 'description': 'mue...
1115                                                   []
1116                                                   []
1117    [{'barcode': 'B076FJ92M4', 'description': 'mue...
1118                                                   []
Name: rewardsReceiptItemList, Length: 1119, dtype: object


In [117]:
#Cross-verifying the number of missing values in the column to ensure the missing values are handled successfully 
missing_value_count__ = receipts_df['rewardsReceiptItemList'].isnull().sum()
print("Number of missing values in 'rewardsReceiptItemList':", missing_value_count__)

Number of missing values in 'rewardsReceiptItemList': 0


Handling the 435 missing values in the 'totalSpent' column.

In [120]:
print(receipts_df['totalSpent'])

0       26.00
1       11.00
2       10.00
3       28.00
4        1.00
        ...  
1114    34.96
1115      NaN
1116      NaN
1117    34.96
1118      NaN
Name: totalSpent, Length: 1119, dtype: float64


In [121]:
# Filling missing values with 0.0
receipts_df['totalSpent'].fillna(0.0, inplace=True)

In [122]:
print(receipts_df['totalSpent'])

0       26.00
1       11.00
2       10.00
3       28.00
4        1.00
        ...  
1114    34.96
1115     0.00
1116     0.00
1117    34.96
1118     0.00
Name: totalSpent, Length: 1119, dtype: float64


In [123]:
#Cross-verifying the number of missing values in the column to ensure the missing values are handled successfully 
_missing_val_count__ = receipts_df['totalSpent'].isnull().sum()
print("Number of missing values in 'totalSpent':", _missing_val_count__)

Number of missing values in 'totalSpent': 0


Changing the values in the 'createDate', 'dateScanned', 'modifyDate' columns

In [129]:
# Defining a custom function to extract the timestamp from the nested dictionary in the 'createDate' column
def extract_timestamp(val):
    if pd.notna(val):
        return val['$date']
    else:
        return val

# Converting the values in 'createDate' to a datetime format using the custom function
receipts_df['createDate'] = receipts_df['createDate'].apply(extract_timestamp)
receipts_df['createDate'] = pd.to_datetime(receipts_df['createDate'], unit='ms')

In [131]:
# Defining a custom function to extract the timestamp from the nested dictionary in the 'dateScanned' column
def extract_timestamp(val):
    if pd.notna(val):
        return val['$date']
    else:
        return val

# Converting the values in 'dateScanned' to a datetime format using the custom function
receipts_df['dateScanned'] = receipts_df['dateScanned'].apply(extract_timestamp)
receipts_df['dateScanned'] = pd.to_datetime(receipts_df['dateScanned'], unit='ms')

In [132]:
# Defining a custom function to extract the timestamp from the nested dictionary in the 'modifyDate' column
def extract_timestamp(val):
    if pd.notna(val):
        return val['$date']
    else:
        return val

# Converting the values in 'modifyDate' to a datetime format using the custom function
receipts_df['modifyDate'] = receipts_df['modifyDate'].apply(extract_timestamp)
receipts_df['modifyDate'] = pd.to_datetime(receipts_df['modifyDate'], unit='ms')

In [133]:
print(receipts_df[['createDate', 'dateScanned', 'modifyDate']])

                  createDate             dateScanned              modifyDate
0    2021-01-03 15:25:31.000 2021-01-03 15:25:31.000 2021-01-03 15:25:36.000
1    2021-01-03 15:24:43.000 2021-01-03 15:24:43.000 2021-01-03 15:24:48.000
2    2021-01-03 15:25:37.000 2021-01-03 15:25:37.000 2021-01-03 15:25:42.000
3    2021-01-03 15:25:34.000 2021-01-03 15:25:34.000 2021-01-03 15:25:39.000
4    2021-01-03 15:25:06.000 2021-01-03 15:25:06.000 2021-01-03 15:25:11.000
...                      ...                     ...                     ...
1114 2021-03-01 10:22:27.000 2021-03-01 10:22:27.000 2021-03-01 10:22:28.000
1115 2021-03-01 15:42:41.873 2021-03-01 15:42:41.873 2021-03-01 15:42:41.873
1116 2021-03-01 14:07:37.664 2021-03-01 14:07:37.664 2021-03-01 14:07:37.664
1117 2021-03-01 13:07:28.000 2021-03-01 13:07:28.000 2021-03-01 13:07:29.000
1118 2021-03-01 02:22:34.962 2021-03-01 02:22:34.962 2021-03-01 02:22:34.962

[1119 rows x 3 columns]


Extracting the time stamps from the nested dictionary in the '_id' column and storing it as strings

In [135]:
# Defining a custom function to extract the timestamp from the nested dictionary
def extract_timestamp(val):
    if pd.notna(val):
        return val['$oid']
    else:
        return val

# Storing the extracted time stamps as strings in the _id column
receipts_df['_id'] = receipts_df['_id'].apply(extract_timestamp)

In [136]:
print(receipts_df['_id'])

0       5ff1e1eb0a720f0523000575
1       5ff1e1bb0a720f052300056b
2       5ff1e1f10a720f052300057a
3       5ff1e1ee0a7214ada100056f
4       5ff1e1d20a7214ada1000561
                  ...           
1114    603cc0630a720fde100003e6
1115    603d0b710a720fde1000042a
1116    603cf5290a720fde10000413
1117    603ce7100a7217c72c000405
1118    603c4fea0a7217c72c000389
Name: _id, Length: 1119, dtype: object


Cleaning the 'rewardsReceiptsItemList'column in the receipts_df dataframe

In [195]:
# Defining a custom function to extract the relevant information from the dictionaries
def extract_rewards_info(item_list):
    if isinstance(item_list, list):
        # If the 'item_list' is a list of dictionaries, extract the desired information
        barcode_list = [item.get('barcode', None) for item in item_list]
        description_list = [item.get('description', None) for item in item_list]
        return barcode_list, description_list
    else:
        # If the 'item_list' is not in the expected format, return None for both lists
        return None, None

# Applying the custom function to the 'rewardsReceiptItemList' column and create new columns for the extracted data
receipts_df[['barcode_list', 'description_list']] = receipts_df['rewardsReceiptItemList'].apply(extract_rewards_info).apply(pd.Series)

# Dropping the original 'rewardsReceiptItemList' column if necessary
receipts_df.drop(columns=['rewardsReceiptItemList'], inplace=True)

In [206]:
receipts_df

_id  bonusPointsEarned  \
0     5ff1e1eb0a720f0523000575              500.0   
1     5ff1e1bb0a720f052300056b              150.0   
2     5ff1e1f10a720f052300057a                5.0   
3     5ff1e1ee0a7214ada100056f                5.0   
4     5ff1e1d20a7214ada1000561                5.0   
...                        ...                ...   
1114  603cc0630a720fde100003e6               25.0   
1115  603d0b710a720fde1000042a                0.0   
1116  603cf5290a720fde10000413                0.0   
1117  603ce7100a7217c72c000405               25.0   
1118  603c4fea0a7217c72c000389                0.0   

                                bonusPointsEarnedReason  \
0     Receipt number 2 completed, bonus point schedu...   
1     Receipt number 5 completed, bonus point schedu...   
2                            All-receipts receipt bonus   
3                            All-receipts receipt bonus   
4                            All-receipts receipt bonus   
...                                                 ...   
1114                        COMPLETE_NONPARTNER_RECEIPT   
1115                                            Unknown   
1116                                            Unknown   
1117                        COMPLETE_NONPARTNER_RECEIPT   
1118                                            Unknown   

                  createDate             dateScanned         finishedDate  \
0    2021-01-03 15:25:31.000 2021-01-03 15:25:31.000  2021-01-03 15:25:31   
1    2021-01-03 15:24:43.000 2021-01-03 15:24:43.000  2021-01-03 15:24:43   
2    2021-01-03 15:25:37.000 2021-01-03 15:25:37.000              Unknown   
3    2021-01-03 15:25:34.000 2021-01-03 15:25:34.000  2021-01-03 15:25:34   
4    2021-01-03 15:25:06.000 2021-01-03 15:25:06.000  2021-01-03 15:25:11   
...                      ...                     ...                  ...   
1114 2021-03-01 10:22:27.000 2021-03-01 10:22:27.000              Unknown   
1115 2021-03-01 15:42:41.873 2021-03-01 15:42:41.873              Unknown   
1116 2021-03-01 14:07:37.664 2021-03-01 14:07:37.664              Unknown   
1117 2021-03-01 13:07:28.000 2021-03-01 13:07:28.000              Unknown   
1118 2021-03-01 02:22:34.962 2021-03-01 02:22:34.962              Unknown   

                  modifyDate    pointsAwardedDate  pointsEarned  \
0    2021-01-03 15:25:36.000  2021-01-03 15:25:31         500.0   
1    2021-01-03 15:24:48.000  2021-01-03 15:24:43         150.0   
2    2021-01-03 15:25:42.000              Unknown           5.0   
3    2021-01-03 15:25:39.000  2021-01-03 15:25:34           5.0   
4    2021-01-03 15:25:11.000  2021-01-03 15:25:06           5.0   
...                      ...                  ...           ...   
1114 2021-03-01 10:22:28.000              Unknown          25.0   
1115 2021-03-01 15:42:41.873              Unknown           0.0   
1116 2021-03-01 14:07:37.664              Unknown           0.0   
1117 2021-03-01 13:07:29.000              Unknown          25.0   
1118 2021-03-01 02:22:34.962              Unknown           0.0   

             purchaseDate  purchasedItemCount rewardsReceiptStatus  \
0     2021-01-03 00:00:00                 5.0             FINISHED   
1     2021-01-02 15:24:43                 2.0             FINISHED   
2     2021-01-03 00:00:00                 1.0             REJECTED   
3     2021-01-03 00:00:00                 4.0             FINISHED   
4     2021-01-02 15:25:06                 2.0             FINISHED   
...                   ...                 ...                  ...   
1114  2020-08-17 00:00:00                 2.0             REJECTED   
1115              Unknown                 0.0            SUBMITTED   
1116              Unknown                 0.0            SUBMITTED   
1117  2020-08-17 00:00:00                 2.0             REJECTED   
1118              Unknown                 0.0            SUBMITTED   

      totalSpent                    userId              barcode_list  \
0          26.00  5ff1e1eacfcf6c399c27

Cleaning the barcode_list and description_list columns in the receipts_df dataframe

In [222]:
# Creating a copy of the DataFrame to avoid modifying the original data
receipts_df_copy = receipts_df.copy()

# Looping through each row in the DataFrame
for index, row in receipts_df_copy.iterrows():
    barcode_list = row['barcode_list']
    if barcode_list:  # Checking if the barcode_list is not empty
        # Converting each element in the barcode_list to integers if they are digits, otherwise keep them as they are
        barcode_list = [int(barcode) if isinstance(barcode, str) and barcode.isdigit() else barcode for barcode in barcode_list]
        receipts_df_copy.at[index, 'barcode_list'] = barcode_list

In [233]:
# Exploding the 'barcode_list' column so that each element in the lists gets its own row
receipts_df_copy_exploded = receipts_df_copy.explode('barcode_list', ignore_index=True)

# Dropping the rows with empty strings in the 'barcode_list'
receipts_df_copy_exploded = receipts_df_copy_exploded[receipts_df_copy_exploded['barcode_list'] != '']

# Resetting the index of the DataFrame
receipts_df_copy_exploded.reset_index(drop=True, inplace=True)

In [228]:
print(receipts_df_copy_exploded['barcode_list'])

0              4011
1              4011
2       28400642255
3               NaN
4              4011
           ...     
3611            NaN
3612            NaN
3613     B076FJ92M4
3614     B07BRRLSVC
3615            NaN
Name: barcode_list, Length: 3616, dtype: object


In [229]:
# Filling NaN values in the 'barcode_list' column with 0
receipts_df_copy_exploded['barcode_list'].fillna(0, inplace=True)

In [230]:
receipts_df_copy_exploded['barcode_list']

0              4011
1              4011
2       28400642255
3                 0
4              4011
           ...     
3611              0
3612              0
3613     B076FJ92M4
3614     B07BRRLSVC
3615              0
Name: barcode_list, Length: 3616, dtype: object

In [234]:
# Exploding the 'description_list' column so that each element in the lists gets its own row
receipts_df_copy_exploded = receipts_df_copy_exploded.explode('description_list', ignore_index=True)

# Dropping the rows with empty strings in the 'description_list'
receipts_df_copy_exploded = receipts_df_copy_exploded[receipts_df_copy_exploded['description_list'] != '']

# Resetting the index of the DataFrame
receipts_df_copy_exploded.reset_index(drop=True, inplace=True)

In [235]:
print(receipts_df_copy_exploded['description_list'])

0                                            ITEM NOT FOUND
1                                            ITEM NOT FOUND
2         DORITOS TORTILLA CHIP SPICY SWEET CHILI REDUCE...
3                                            ITEM NOT FOUND
4         DORITOS TORTILLA CHIP SPICY SWEET CHILI REDUCE...
                                ...                        
378870    mueller austria hypergrind precision electric ...
378871    thindust summer face mask - sun protection nec...
378872    mueller austria hypergrind precision electric ...
378873    thindust summer face mask - sun protection nec...
378874                                                  NaN
Name: description_list, Length: 378875, dtype: object


In [239]:
# Filling NaN values in the 'description_list' column with 'unknown'
receipts_df_copy_exploded['description_list'].fillna('unknown', inplace=True)

In [237]:
print(receipts_df_copy_exploded['description_list'])

0                                            ITEM NOT FOUND
1                                            ITEM NOT FOUND
2         DORITOS TORTILLA CHIP SPICY SWEET CHILI REDUCE...
3                                            ITEM NOT FOUND
4         DORITOS TORTILLA CHIP SPICY SWEET CHILI REDUCE...
                                ...                        
378870    mueller austria hypergrind precision electric ...
378871    thindust summer face mask - sun protection nec...
378872    mueller austria hypergrind precision electric ...
378873    thindust summer face mask - sun protection nec...
378874                                                    0
Name: description_list, Length: 378875, dtype: object


In [242]:
# Replacing 0 values with 'unknown'
receipts_df_copy_exploded['description_list'].replace(0, 'Unknown', inplace=True)

In [243]:
print(receipts_df_copy_exploded['description_list'])

0                                            ITEM NOT FOUND
1                                            ITEM NOT FOUND
2         DORITOS TORTILLA CHIP SPICY SWEET CHILI REDUCE...
3                                            ITEM NOT FOUND
4         DORITOS TORTILLA CHIP SPICY SWEET CHILI REDUCE...
                                ...                        
378870    mueller austria hypergrind precision electric ...
378871    thindust summer face mask - sun protection nec...
378872    mueller austria hypergrind precision electric ...
378873    thindust summer face mask - sun protection nec...
378874                                              unknown
Name: description_list, Length: 378875, dtype: object


In [254]:
# Assuming 'receipts_df' is your DataFrame
receipts_df_copy = receipts_df.copy()

# Exploding the 'barcode_list' and 'description_list' columns
receipts_df_copy_exploded = receipts_df_copy.explode('barcode_list', ignore_index=True)
receipts_df_copy_exploded['description_list'] = receipts_df_copy['description_list'].explode(ignore_index=True)

# Dropping rows with empty values in 'barcode_list' and 'description_list'
receipts_df_copy_exploded = receipts_df_copy_exploded.dropna(subset=['barcode_list', 'description_list'])

# Resetting the index of the DataFrame
receipts_df_copy_exploded.reset_index(drop=True, inplace=True)

In [271]:
receipts_df_copy_exploded

_id  bonusPointsEarned  \
0     5ff1e1eb0a720f0523000575              500.0   
1     5ff1e1bb0a720f052300056b              150.0   
2     5ff1e1bb0a720f052300056b              150.0   
3     5ff1e1ee0a7214ada100056f                5.0   
4     5ff1e1d20a7214ada1000561                5.0   
...                        ...                ...   
2846  603cc2bc0a720fde100003e9               25.0   
2847  603cc0630a720fde100003e6               25.0   
2848  603cc0630a720fde100003e6               25.0   
2849  603ce7100a7217c72c000405               25.0   
2850  603ce7100a7217c72c000405               25.0   

                                bonusPointsEarnedReason          createDate  \
0     Receipt number 2 completed, bonus point schedu... 2021-01-03 15:25:31   
1     Receipt number 5 completed, bonus point schedu... 2021-01-03 15:24:43   
2     Receipt number 5 completed, bonus point schedu... 2021-01-03 15:24:43   
3                            All-receipts receipt bonus 2021-01-03 15:25:34   
4                            All-receipts receipt bonus 2021-01-03 15:25:06   
...                                                 ...                 ...   
2846                        COMPLETE_NONPARTNER_RECEIPT 2021-03-01 10:32:28   
2847                        COMPLETE_NONPARTNER_RECEIPT 2021-03-01 10:22:27   
2848                        COMPLETE_NONPARTNER_RECEIPT 2021-03-01 10:22:27   
2849                        COMPLETE_NONPARTNER_RECEIPT 2021-03-01 13:07:28   
2850                        COMPLETE_NONPARTNER_RECEIPT 2021-03-01 13:07:28   

             dateScanned         finishedDate          modifyDate  \
0    2021-01-03 15:25:31  2021-01-03 15:25:31 2021-01-03 15:25:36   
1    2021-01-03 15:24:43  2021-01-03 15:24:43 2021-01-03 15:24:48   
2    2021-01-03 15:24:43  2021-01-03 15:24:43 2021-01-03 15:24:48   
3    2021-01-03 15:25:34  2021-01-03 15:25:34 2021-01-03 15:25:39   
4    2021-01-03 15:25:06  2021-01-03 15:25:11 2021-01-03 15:25:11   
...                  ...                  ...                 ...   
2846 2021-03-01 10:32:28              Unknown 2021-03-01 10:32:29   
2847 2021-03-01 10:22:27              Unknown 2021-03-01 10:22:28   
2848 2021-03-01 10:22:27              Unknown 2021-03-01 10:22:28   
2849 2021-03-01 13:07:28              Unknown 2021-03-01 13:07:29   
2850 2021-03-01 13:07:28              Unknown 2021-03-01 13:07:29   

        pointsAwardedDate  pointsEarned         purchaseDate  \
0     2021-01-03 15:25:31         500.0  2021-01-03 00:00:00   
1     2021-01-03 15:24:43         150.0  2021-01-02 15:24:43   
2     2021-01-03 15:24:43         150.0  2021-01-02 15:24:43   
3     2021-01-03 15:25:34           5.0  2021-01-03 00:00:00   
4     2021-01-03 15:25:06           5.0  2021-01-02 15:25:06   
...                   ...           ...                  ...   
2846              Unknown          25.0  2020-08-17 00:00:00   
2847              Unknown          25.0  2020-08-17 00:00:00   
2848              Unknown          25.0  2020-08-17 00:00:00   
2849              Unknown          25.0  2020-08-17 00:00:00   
2850              Unknown          25.0  2020-08-17 00:00:00   

      purchasedItemCount rewardsReceiptStatus  totalSpent  \
0                    5.0             FINISHED       26.00   
1                    2.0             FINISHED       11.00   
2                    2.0             FINISHED       11.00   
3                    4.0             FINISHED       28.00   
4                    2.0             FINISHED        1.00   
...                  ...                  ...         ...   
2846                 2.0             REJECTED       34.96   
2847                 2.0             REJECTED       34.96   
2848                 2.0             REJECTED       34.96   
2849                 2.0             REJECTED       34.96   
2850                 2.0             REJECTED       34.96   

                        userId barcode_list  \
0     5ff1e1eacfcf6c399c274ae6         4011   
1     5ff1e194b6

Saving the modified dataframes to new CSV files

In [261]:
# Saving the modified User DataFrame to a new CSV file called 'modified_users.csv'
users_df.to_csv('modified_users.csv', index=False)
# Saving the modified Brands DataFrame to a new CSV file called 'modified_brands.csv'
brands_df.to_csv('modified_brands.csv', index=False)
# Saving the modified Receipts DataFrame to a new CSV file called 'modified_receipts.csv'
receipts_df_copy_exploded.to_csv('modified_recipts.csv', index=False)